### PR TITLE
simplify rollup setup

### LINKED ISSUE #2108: Simplify rollup config and submodules
Currently, when we define a new export in the packages we add it in several places:

**Example** for `@trpc/server/observables`:

- https://github.com/trpc/trpc/blob/8c5499cb42697db3c0b580e92aa176cfef5b2475/packages/server/rollup.config.js#L14
- https://github.com/trpc/trpc/blob/8c5499cb42697db3c0b580e92aa176cfef5b2475/packages/server/package.json#L82-L86
- https://github.com/trpc/trpc/blob/8c5499cb42697db3c0b580e92aa176cfef5b2475/packages/server/package.json#L98
- https://github.com/trpc/trpc/blob/8c5499cb42697db3c0b580e92aa176cfef5b2475/packages/server/observable/index.js#L1
- https://github.com/trpc/trpc/blob/8c5499cb42697db3c0b580e92aa176cfef5b2475/packages/server/test/___packages.ts#L4-L5

This is very fiddly and easy to get wrong. It would be nice if we could automate most of the parts of this.

I did a first attempt of this with trying to automate the rollup stuff in https://github.com/trpc/trpc/blob/next/scripts/rollup.js, but it's a terrible hacky script & currently only works for `@trpc/client` and doesn't cover edge-cases enough to work for `/react`, `/next`, `/server`.

If someone likes build-tools and has some good ideas, feel free to submit a PR to the `next` branch with automating this as much as possible. I do still miss preconstruct which we have on `main` which automatically did this for us.

### PR DESCRIPTION
**Pull request by juliusmarminge**
Redo of #2376 

Took two steps back for now and only simplified the rollup config.

Also rewrote the entrypoints script. It's now way simpler and works for all packages. It syncs the `exports` field in the package.json and creates the necessary barrel-files.

Personally, I wouldn't mind adding the barrels to a .gitignore so that the repo is cleaner to browse? (yes, I go to the barrelfile instead of the sourcefile way too often...)

Closes #2108

### CODE REVIEW

File: packages/server/adapters/aws-lambda/index.d.ts [hunk 1]
```diff
@@ -1 +1 @@
-export * from '../../dist/adapters/aws-lambda';
+export * from '../../dist/adapters/aws-lambda/index';
```
Comment (juliusmarminge) [inline packages/server/adapters/aws-lambda/index.d.ts:1]: could escape the index part but it doesn't really hurt so keeping the script as minimal as possible

**Comment by juliusmarminge** [comment]
It should be possible to automate this even further, and use the script for the exports field in package.json too from the inputs in the rollup config.

File: packages/client/README.md [hunk 2]
```diff
@@ -1,60 +0,0 @@
...
-  // Querying the greeting
-  const helloResponse = await proxy.greeting.query({
-    name: 'world',
-  });
-
-  console.log('helloResponse', helloResponse); // Hello world
-}
-
-main();
-```
```
Comment (KATT) [inline packages/client/README.md:60]: Is this intended..?

**Review by KATT — COMMENTED** [review]
Questions & comments!

- Overall looks awesome and makes the whole thing of adding a new entrypoint a lot less error prone
- Happy to move barell files to gitignore
- Will `yarn dev` just work or does it require a `build` first`? If so, please make the `postinstall` script make it "just work"
- How come `README.md` and `LICENSE` files were removed?
- @sachinraja WDYT?

**Comment by juliusmarminge** [comment]
> * Will `yarn dev` just work or does it require a `build` first`? If so, please make the `postinstall` script make it "just work"

Will look into the scripts tomorrow. I believe there are a few that needs modification.

> * How come `README.md` and `LICENSE` files were removed?

Sorry bout that. Removed some stuff as I got things working. Will restore those 👍🏼

**Comment by juliusmarminge** [comment]
Just checked and following the contributing guide works, all you need is a `yarn` command which will install deps and do initial build and codegen, then `yarn dev` will startup watch-mode 👍🏼 

@sachinraja one thing I don't really understand is it includes `test` and `examples` on build, even though they are excluded in the tsconfigs, any idea why?
CleanShot 2022-08-09 at 13 35 22@2x

**Comment by juliusmarminge** [comment]
Whoops, looks like the CI didn't like moving jest-config to root (the examples are picking it up it seems like. Should I move it back or modify it so that only the packages uses it? Even though all tests are in server right now, to me, it feels more intuitive to have a single config for all packages if we were to add tests to the other ones as well down the road.

**Comment by juliusmarminge** [comment]
Rewrote the script again, now it is fully automated.

Put a new entry in the rollup.config.ts INPUTS, and the barrelfile will be created, the `exports` defined in package.json as well as the "top-level-barrel-directory" will be added to `files`.

File: .gitignore [hunk 3]
```diff
@@ -12,4 +12,13 @@ yalc.lock
...
+*.tsbuildinfo
+
+# Barrel files
+packages/client/links
+packages/react/ssg
+packages/server/adapters
+packages/server/observable
+packages/server/rpc
+packages/server/shared
+packages/server/subscription
```
Comment (sachinraja) [inline .gitignore:24]: I imagine these will get annoying to ignore as it's not automated, I'm fine with just committing them anyway.

File: package.json [hunk 4]
```diff
@@ -37,23 +37,26 @@
   "scripts": {
     "lerna": "lerna",
     "start": "yarn watch",
+    "clean:packages": "lerna run clean --scope=@trpc/*",
+    "clean:examples": "lerna run clean --scope=@examples/*",
+    "clean:root": "rm -rf node_modules",
+    "clean": "run-p clean:*",
```
Comment (sachinraja) [inline package.json:43]:
```suggestion
    "clean:packages": "rm -rf **/node_modules **/dist",
    "clean:examples": "lerna run clean --scope=@examples/*",
    "clean": "run-p clean:*",
```
This way we won't need the `clean` scripts in each `package.json` and also won't need to have `node_modules` in each of the clean scripts for the examples.

File: packages/client/README.md [hunk 5]
```diff
@@ -1,60 +1,52 @@
 <p align="center">
```
Comment (sachinraja) [inline packages/client/README.md:1]: Can we revert the changes here? Unsure what the large diff is for.

File: packages/client/LICENSE [hunk 6]
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Comment (sachinraja) [inline packages/client/LICENSE:1]: Should this be here?

File: packages/next/LICENSE [hunk 7]
```diff
@@ -1,21 +1,21 @@
 MIT License
```
Comment (sachinraja) [inline packages/next/LICENSE:1]: We should probably delete all these individual licenses and just have one in the root.

File: package.json [hunk 8]
```diff
@@ -80,11 +83,14 @@
     "@changesets/cli": "^2.17.0",
     "@manypkg/cli": "^0.19.0",
     "@rollup/plugin-babel": "^5.3.0",
+    "@rollup/plugin-multi-entry": "^4.1.0",
     "@rollup/plugin-node-resolve": "^13.1.3",
+    "@rollup/plugin-typescript": "^8.3.4",
```
Comment (sachinraja) [inline package.json:88]: Since we're using `rollup-plugin-typescript2`, we should be able to uninstall this.

File: rollup.config.ts [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+const extensions = ['.ts', '.tsx'];
+
+const babelPlugin = babel({
+  babelHelpers: 'runtime',
+  extensions,
+});
+
+// Exporting this for generating barrel-files in scripts/entrypoints.ts
+export const PACKAGES = ['server', 'client', 'react', 'next'] as const;
+export const INPUTS: Record<typeof PACKAGES[number], string[]> = {
```
Comment (sachinraja) [inline rollup.config.ts:24]: Love that this is strictly typed :smile:

File: rollup.config.ts [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+  return [types(options), esm(options), cjs(options)];
+}
+
+function types({ input, packageDir }: Options): RollupOptions {
+  return {
+    input,
+    output: {
+      dir: `${packageDir}/dist`,
+      entryFileNames: '[name].ts.js',
+      chunkFileNames: '[name]-[hash].ts.js',
```
Comment (sachinraja) [inline rollup.config.ts:95]: Just as a note if you're confused as to why I had this here, it's because `rollup-plugin-typescript2` produces JS files that'll overwrite the files from the ESM build... If you have a better idea or know of a better rollup plugin/TS integration that would be awesome.

File: rollup.config.ts [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+      externals({
+        packagePath: path.resolve(packageDir, 'package.json'),
+      }),
+      nodeResolve({
+        extensions,
+      }),
+      babelPlugin,
+    ],
+  };
+}
```
Comment (sachinraja) [inline rollup.config.ts:155]: Nice!

File: scripts/entrypoints.ts [hunk 12]
```diff
@@ -1,71 +1,103 @@
...
 
-function getEntrypoint(filename: string) {
-  const parts = filename.split('/');
-  const name = parts.pop();
+// create directories on the way if they dont exist
+function writeFileSyncRecursive(filePath: string, content: string) {
+  const dir = path.dirname(filePath);
+  if (!fs.existsSync(dir)) {
+    fs.mkdirSync(dir, { recursive: true });
+  }
```
Comment (sachinraja) [inline scripts/entrypoints.ts:18]:
```suggestion
fs.mkdirSync(dir, { recursive: true, force: true });
```
nit

**Review by sachinraja — CHANGES_REQUESTED** [review]
> it feels more intuitive to have a single config for all packages if we were to add tests to the other ones as well down the road.

I agree, but let's leave it in `packages/server` for this PR at least.

Nice job on this! This makes changing entry points so much easier.

File: .gitignore (hunk 3 above)
Comment (juliusmarminge) [reply .gitignore:24]: Yea I initially wanted a glob pattern here to say like, ignore subdirectory of `packages/*` but `src` and `test` but couldn't get that to work. You know what pattern to use?

File: package.json (hunk 4 above)
Comment (juliusmarminge) [reply package.json:43]: Yep I believe I had something like that earlier. However, I quite like having a clean command for everyone so that when I am working on something and want to make sure nothing is "left stale", I run a `clean` for that project only. Can definitely change to your suggestion if you don't like this.

File: packages/client/README.md (hunk 5 above)
Comment (juliusmarminge) [reply packages/client/README.md:1]: Absolutely, not sure why it's diffing entire file, the content is the exact same...

File: packages/client/LICENSE (hunk 6 above)
Comment (juliusmarminge) [reply packages/client/LICENSE:1]: There were `LICENSE` in some of the packages earlier so I just added it to all of them. I also think it's enough to have a single one in root if @KATT is fine with that too?

File: scripts/entrypoints.ts (hunk 12 above)
Comment (sachinraja) [reply scripts/entrypoints.ts:18]: nevermind, apparently I made this option up, it doesn't exist :laughing:

File: rollup.config.ts (hunk 10 above)
Comment (juliusmarminge) [reply rollup.config.ts:95]: Yea we need to overwrite the defaults for these fields so that they don't overwrite eachother... tried to get the typescript one to only emit the declaration files but that didn't seem to do it. Can take a second look and see if there isn't a way 👍

File: package.json (hunk 8 above)
Comment (juliusmarminge) [reply package.json:87]: Actually, I think we do need this. This plugin is the one which allows us to have a typescript config.

File: rollup.config.ts (hunk 10 above)
Comment (juliusmarminge) [reply rollup.config.ts]: > tried to get the typescript one to only emit the declaration files but that didn't seem to do it

Looks like it did do it... no more empty files :)

We do need these options for the ESM output though, could remove for the CJS since that is using the defaults but better to be explicit about the differences there imo.

File: package.json (hunk 8 above)
Comment (sachinraja) [reply package.json:87]: Ah I see

File: .gitignore [hunk 13]
```diff
@@ -12,4 +12,10 @@ yalc.lock
...
 
-*.tsbuildinfo
\ No newline at end of file
+*.tsbuildinfo
+
+# Barrel files
+packages/*/*
+!packages/*/package.json
+!packages/*/src
+!pacakges/*/README.md
```
Comment (KATT) [inline .gitignore:21]:
```suggestion
!packages/*/README.md
```

Comment (KATT) [inline .gitignore]: Could this work?

```suggestion
# Barrel files
## Ignore all folders
packages/*/*
## Don't ignore files
!packages/*/*.*
## Don't ignore `src` folder
!packages/*/src
```

File: examples/cloudflare-workers/package.json [hunk 14]
```diff
@@ -5,6 +5,7 @@
   "type": "module",
   "scripts": {
     "dev": "wrangler dev",
+    "clean": "rm -rf dist node_modules",
```
Comment (KATT) [inline examples/cloudflare-workers/package.json:8]: Should we get rid of these clean scripts? They only work on windows and I don't know why we need them


Instead, we should have a `clean` file in the root that does this if/when it's needed

Comment (juliusmarminge) [reply examples/cloudflare-workers/package.json:8]: Could absolutely remove them if you don't like them. I like having clean scripts if I want to test something from a clean sleight.

File: .gitignore (hunk 13 above)
Comment (juliusmarminge) [reply .gitignore]: Yep that works. Also included `test` which was missing 👍🏼

File: examples/cloudflare-workers/package.json (hunk 14 above)
Comment (KATT) [reply examples/cloudflare-workers/package.json:8]: I like having them too but the only example I really care about being representative of a "production environment" is the `next-prisma-starter`, the otherwise shouldn't contain unnecessary noise

Comment (juliusmarminge) [reply examples/cloudflare-workers/package.json:8]: Alright, I'll remove them

**Review by sachinraja — APPROVED** [review]
